marimo/_save/ast.py

```diff
@@ -106,31 +112,69 @@ def generic_visit(self, node: ast.AST) -> tuple[ast.Module, ast.Module]:  #  typ
         # excluding by omission try, for, classes and functions.
         if len(on_line) == 0:
             if isinstance(previous, (ast.With, ast.If)):
-                try:
-                    # Recursion by referring the the containing block also
-                    # captures the case where the target line number was not
-                    # exactly hit.
-                    return ExtractWithBlock(self.target_line).generic_visit(
-                        previous.body  # type: ignore[arg-type]
-                    )
-                except BlockException:
-                    on_line.append(previous)
+                on_line.append(previous)
+                bodies = (
+                    [previous.body]
+                    if isinstance(previous, ast.With)
+                    else [previous.body, previous.orelse]
+                )
+                for body in bodies:
+                    try:
+                        # Recursion by referring the the containing block also
+                        # captures the case where the target line number was not
+                        # exactly hit.
+                        return ExtractWithBlock(
+                            self.target_line
+                        ).generic_visit(
+                            body  # type: ignore[arg-type]
+                        )
+                    except BlockException:
+                        pass
+
             else:
                 raise BlockException(
                     "persistent_cache cannot be invoked within a block "
                     "(try moving the block within the persistent_cache scope)."
                 )
         # Intentionally not elif (on_line can be added in previous block)
         if len(on_line) == 1:
-            assert isinstance(on_line[0], ast.With), "Unexpected block."
+            if parent and not isinstance(on_line[0], ast.With):
+                raise BlockException("Detected line is not a With statement.")
+            if not isinstance(on_line[0], ast.With):
+                raise BlockException(
+                    "Saving on a shared line may lead to unexpected behavior. "
```

Comment:
Can you make this error more descriptive? I think our users may not understand what "Saving on a shared line" means, or how to fix it.

Perhaps an example of an offending code block and how to fix it?